### ISSUE TITLE
Adding kerning cases breaks previously-working ones

### ISSUE THREAD

**Issue by bobh0303** [silnrsi/font-lateef#2]
## Description

When I gradually add more kerning cases, ones that previously kerned correctly no longer do so.

## Details

There are 10 alef-like characters in the font:

| |Working name|Production name|USV|
|-|------------|---------------|---|
|1|alefMadda-ar|uni0622|0622|
|2|alefHamzaabove-ar|uni0623|0623|
|3|alefHamzabelow-ar|uni0625|0625|
|4|alef-ar|uni0627|0627|
|5|alefWasla-ar|uni0671|0671|
|6|alefWavyhamzaabove-ar|uni0672|0672|
|7|alefWavyhamzabelow-ar|uni0673|0673|
|8|highhamzaAlef-ar|uni0675|0675|
|9|alefTwoabove-ar|uni0773|0773|
|10|alefThreeabove-ar|uni0774|0774|

The goal is to build, using grkern2fea, OpenType kerning for U+0631 REH (isolate and final) followed by any one of the alef-like characters, accounting for needed combinations of intervening and following marks.

If `source/kerndata.ftml` (input to grkern2fea) includes records for just first 4 alefs, the kerning works fine. When the 5th alef is added the kerning builds but doesn't work.

## To reproduce:

These commits on the `collision-based-kerning` branch are relevant:
|SHA|description|
|--|-----------|
| e786ea4a | kerndata from just reh and 4 alefs -- this seems to work (except for expected issues with three diac above alef)|
| b61e7150 | kerndata from just reh and 5 alefs -- this does not seem to work|

1. Check out either of those two commits
2. Optional: run `tools/updateKerning.sh` (no arguments needed) to rebuild the `*-caKern.fea` files. NB: The .fea files generated on my VM from this tool are included in the commit, but you may want to rebuild them to see if you're getting the same thing.
3. Build the fonts, being sure to add `ftml` to the targets and, if desired, `--norename` to prevent glyph renaming:
```
smith distclean
smith configure
smith build ftml --noRename
```
4. To test, use the `kerntest-short.ftml` file. NB: See the Rendering Check to verify that data columns 2 and 4 are actually rendering OpenType, not Graphite.
5. Repeat with the other commit

## Observations

In `kerntest-short.ftml` the 4-alef case kerns nicely:
![image](https://user-images.githubusercontent.com/6842108/133806456-9592b434-34bd-462c-b285-2e29b7882ef2.png)
but the 5-alef case does not:
![image](https://user-images.githubusercontent.com/6842108/133806617-8df5cf9c-7276-4e1c-9bd6-05e0feab8c24.png)

Looking at the `Lateef-Regular-caKern.fea` files from the two cases, one difference I notice is that in the working case the generated fea is "single stage" with just one contextual lookup (`mainkern`), whereas in the failing case it is two-stage, with the `mainkern` contextual lookup sometimes calling a second contextual lookup (`kernposchain_0`).

**Comment by bobh0303** [comment]
Attached zip includes `kerntest-short.ftml`

[kerntest-short.zip](https://github.com/silnrsi/font-lateef/files/7186844/kerntest-short.zip)

(sorry, but github doesn't support attaching ftml files ;-)

**Comment by bobh0303** [comment]
Thanks for silnrsi/palaso-python@ce8cceb0392c7662d13fcfd9ec6c068a4fca47f8, and I see additional fea rules now being generated.

However, we're not there yet. Consider the simple sequence `reh`+`alef`+`kasratan`, i.e. 0631,0627,064D. The raw pair data:
```
"راٍ" [reh-ar]{243.506}!0 [alef-ar]!1 [kasratan-ar]!2
```
suggests the reh needs to be kerned away from the alef when alef has a kasratan under it. However what we see in the resulting rendering is: 
![image](https://user-images.githubusercontent.com/6842108/134446049-1568f4c8-6979-4db0-bb09-ecbe98647a43.png)

**Comment by bobh0303** [comment]
Actually, I'm quite confused by `mainkern`. The first three rules are:
```
@kernClass_1 = [reh-ar reh-ar.fina];  # used 36 times
@kernClass_2 = [alef-ar alefWasla-ar];  # used 17 times
@kernClass_3 = [fathatan-ar kasratan-ar];  # used 6 timeslookup mainkern {
...
lookup mainkern {
    pos reh-ar' lookup kernpos_0 kasratan-ar @kernClass_2;
    pos @kernClass_1' lookup kernpos_2 fathatan-ar @kernClass_2;
    pos @kernClass_1' lookup kernpos_1 @kernClass_2;
```
With the current [limited] set of input data, these three match *all* the raw pair data, regardless of what follows the `@kernClass_2` (i.e., no matter what marks follow the alef or alef-wasla) -- so how do any of the other rules in `mainkern` ever contribute anything?  I would have thought these would be the *last* three rules in the lookup.

What am I missing?

**Comment by mhosken** [comment]
You have missed nothing. I have committed new code to grkern2fea (well glyphstring.py in palaso-python) that checks the overlapping better and now moves those 3 rules to after the other rules as  a result. Might this work?

**Comment by bobh0303** [comment]
Thanks for silnrsi/palaso-python@e399261e6656234870a2ebd60bc5eb9d373f0d14. Checking...

**Comment by bobh0303** [comment]
Still not working.  I've committed the new caKern.fea files to the branch. 

Take the two sequences:
* reh + fathatan + alef                     <0631,064B,0627> 
* reh + fathatan + alef + kasratan   <0631,064B,0627,064D>

which should look like the Graphite-rendered highlighted sequences below:
![image](https://user-images.githubusercontent.com/6842108/135022029-7e8a087c-b52f-4e28-b923-ee3f5d33121d.png)

but in fact look like:
![image](https://user-images.githubusercontent.com/6842108/135021869-6cf4d266-05a4-466a-84ff-ba375b0793dd.png)

The kasratan below the alef should cause the reh to move right but doesn't.

Studying [rawPairData--Regular-nozwj.txt](https://github.com/silnrsi/font-lateef/files/7241138/rawPairData--Regular-nozwj.txt) and generated [Regular-caKern.fea](https://github.com/silnrsi/font-lateef/blob/collision-based-kerning/source/opentype/Lateef-Regular-caKern.fea) :

The first sequence:
* has graphite kern of -46.1294  (rawData line 77)
* matches `mainkern` line 61: `pos @kernClass_1' lookup kernpos_2 fathatan-ar @kernClass_2;`
* in which lookup `kernpos_2` correctly kerns the reh  `<-44 0 -44 0>`

However, the second sequence:
* has graphite kern of 243.506  (rawData line 59)
* matches `mainkern` line 52: `pos @kernClass_1' lookup kernposchain_0 @kernClass_3' @kernClass_2' kasratan-ar;`
* which, in `kernposchain_0` matches line 39: `pos @kernClass_1' lookup kernpos_2 fathatan-ar @kernClass_2;`
* which uses the same lookup `kernpos_2`, thus again kerning the reh `<-44 0 -44 0>` which is incorrect

.

**Comment by bobh0303** [comment]
I don't see how `kernposchain_0` can work at all for the second sequence since it contains no rules that match a string with a diacritic under the alef.

**Comment by bobh0303** [comment]
You asked:
> If `kernpos_2` isn't the right lookup, what one should it have been? 

Based on a desired kern of ~244 it would be either `kernpos_3` or `kernpos_4` I think. 

Of the `kernposchain_0` rules that use either of these lookups, I think the only one that matches the data we're studying is line 41:
```
pos @kernClass_1' lookup kernpos_3 fathatan-ar alef-ar;
```
but, as we have seen, it is being masked by the earlier line 39 that executes `kernpos_2`.

Note that you can't just reverse the order of these rules because then the other would be masked. 

Note also: If we were to try `alefWasla-ar` in our test data instead of `alef-ar` we would get the same problem: line 39 masks the desired rule but in this case the desired rule is line 43:
```
pos @kernClass_1' lookup kernpos_4 fathatan-ar alefWasla-ar;
```

**Comment by mhosken** [comment]
Where can I find this test data so that I can reproduce the problem? and dig out test sequences, etc. Somewhat frustrating. Because my reading of caKern.fea implies to me that all is well. But I can't test it.

**Comment by bobh0303** [comment]
fixed by silnrsi/palaso-python@a4c2178607936211c0993b8e0941c25f8448ac11  Many thanks, Martin!
